### ISSUE TITLE
Add Player::isRandomRace()

### ISSUE THREAD

**Issue by chriscoxe** [bwapi/bwapi#679]
I suggest adding a new method:
bool BWAPI::Player::isRandomRace()
which can be used to query whether a specific player (myself, or an enemy player or ally at least) picked the "Random" race in the game lobby.  For neutral players (e.g. the player that critters belong to), presumably it should return false.  I am not sure whether this proposal is technically feasible or how hard it would be to implement though.  If it is too hard to implement then I wouldn't bother.

Rationale: to be able to adjust my play according to whether myself and/or enemy players and/or allies picked Random race (i.e. regardless of whether any of their units were visible to me at the start of the game).  In a bot that supports race-picking and/or against enemy bot(s) that support race-picking, this would be useful when deciding what strategy and starting build order to use at the start of the game.  I realize that in the CIG/AIIDE/SSCAIT competitions these days every bot has to stick to a particular race and stick with it for the entire competition (Protoss or Zerg or Terran or Random), but in future I foresee competitions like the mini-tournaments organized by Martin Rooijackers where players are allowed to race-pick (e.g. switch between picking Zerg and picking Random depending on the circumstances).  Race-picking could be done based on map (e.g. always play Zerg on Blood Bath map and always play Random race on Andromeda map) and/or based on which opponent they are playing against (e.g. always pick Random against LetaBot, always pick Random against newer versions of UAlbertaBot that pick Random, always pick Zerg against older versions of UAlbertaBot that can only play as Protoss).  http://wiki.teamliquid.net/starcraft/Racepicking also talks about being able to pick your race depending on what race your opponent has picked, but I suppose that wouldn't work if more than one players wants to race-pick.

Note: if everyone picks "Random" race in the game lobby, calling BWAPI::Player::getRace() for myself or my allies (so long as I have seen at least one of their units yet) will just return the race that was rolled (Protoss or Zerg or Terran, not Random), and calling BWAPI::Player::getRace() for an enemy player will return Unknown if I have not seen one of their units yet otherwise their actual race (Protoss or Zerg or Terran, not Random).  That is fine - I am not proposing to change that logic.  So, for enemy players that had no units visible to me at the start of the game, it's already possible to know whether they were Random race.  However, currently it is not possible to know whether I picked Random race or whether enemy players picked Random race that had units visible to me at the start of the game (e.g. presumably this would happen if the Complete Map Information flag is turned on but I have not tried it).

**Comment by ratiotile** [comment]
Does this have something to do with the skewed probabilities of getting a mirror match in a RvR? I don't see any other possible advantage here.

**Comment by chriscoxe** [comment]
It's mainly to pick a strategy at the start of the game.  E.g. suppose my enemy picked Terran, and I picked Zerg or picked Random and happened to roll Zerg.  If I picked Zerg it means that my enemy also knows I picked Zerg.  If they are clever they will realize that they only need to deal with Zerg strategies (they don't need to bother scouting for proxy rax/gateways or cannon rushes etc) and will adapt their play accordingly (e.g. they may be more likely to wall off their ramp because they know I am Zerg).  Alternatively, if I picked Random and happened to roll Zerg, my opponent doesn't know my race until they see one of my units, so if they are clever they will adjust their play (e.g. send out a scout earlier than usual so they are more likely to find out my race earlier than usual) and e.g. may be less likely to wall off since they know I am unlikely to be Zerg.  So if I am race-picking, I could use the proposed method to help me decide whether to use a strategy that is weak against wall-offs (or decide what % to use that strategy if I am using that strategy as part of a mixed strategy).

**Comment by ratiotile** [comment]
I suppose it could be used in the case that you race-pick Random against a certain matchup, but I didn't catch that the first time, because it is almost never done that way. It's usually choosing a specific race because the player has more confidence in that matchup.

Another way to solve this problem would be to add a lobby api and load the bots in the lobby so that custom race selection logic can be used.

**Comment by chriscoxe** [comment]
Yes, although if I am understanding you correctly, to make use of that lobby api it would mean when I release my bot I'd have to build multiple variants of my bot in each release (say, a bot that only plays as Zerg, a bot that only plays as Protoss, a bot that only plays as Terran, a bot that only plays as Random, or some kind of config setting that has a different value for each variant), which I would rather avoid, and I would want the read/write folders would need to be shared by all four variants (rather than separate) which makes things more complicated in BWAPI and/or StarcraftAITournamentManager.

**Comment by ratiotile** [comment]
No, I was thinking of the lobby api as the bot having the ability to see the race of its opponent in the lobby. Each bot would choose their initial race, then each bot has 1 chance to change race after observing the opponent's race. By loading the bot in the lobby, it will know what race it & the opponent has picked going into the game.

**Comment by chriscoxe** [comment]
I see - dynamic race-picking by the bot :)

But it would still not solve the problem of knowing when the game starts whether Random or Terran/Protoss/Zerg was picked.

**Comment by ratiotile** [comment]
Well, by race I mean Random/Protoss/Terran/Zerg, but yeah either each bot would have a final chance to observe it and store it in a variable before the game starts, or as you suggested, bwapi records it and provides a method to get that information. Lobby api would be a major change, but will also handle whatever race-pick variants people can think of without having to build them into bwapi.
